### PR TITLE
RD-1853 Filter by parent deployment in Deployments View

### PR DESCRIPTION
**Pull request by Gelio**
This PR adds filtering by parent deployment in the Deployments View widget. It covers RD-1853

Some of the changes to the core of the application result from the fact that `drilldownContext` only contained entries for each drill-down page, but not the topmost page. This would make it impossible to filter by the deployment selected on the topmost Environments/Services page because the selected deployment would be gone. Now, all pages, no matter if drilled-down, or top-level, are present in the `drilldownContext` array in the store.

As a side effect, when going back to some predecessor page by clicking the breadcrumbs, the context is now restored. If you set some configuration option on the top-level page (e.g. set a filter), drill down, and go back to that top-level page, the filter will be restored. IMO that's a nice UX improvement.

There is a known issue and incompatibility between the _Deployments_ and the _Deployments View_ widget in how they handle the `deploymentId` context. See RD-1984 for more details.

Some of the problems I tackled in this PR could have been solved in RD-1226, but then I would not be able to confidently tell that the functionality of filtering by parent deployment is working :smile: 

## Video and how to test

I found it easiest to test the functionality in the following manner:
1. Have the following structure of pages when drilling-down in the Deployments widget:
    1. Top-level page with Deployments widget.
    2. A drilled-down page, with Deployments widget and with Deployments View widget configured to filter by parent deployment.
    3. A drilled-down page, with Deployments widget and with Deployments View widget configured to filter by parent deployment.
2. On the top-level page, select a deployment that has children. In my video, that's `app-env`
3. In the first drill-down page, notice that there is a warning that there is no parent deployment. This is expected due to RD-1984 (`deploymentId` is set not on the parent page, but on the current page).
4. On that first drill-down page, select any deployment. In my video, that's `db-1` (but it does not matter)
5. In the second drill-down page, notice that the Deployments View widget shows the deployments that are directly connected to `app-env` (the parent deployment, meaning the `deploymentId` set on the parent (first drill-down) page). Those deployments are `db-env` and `web-app`. This is expected.

Additionally, notice that when going to the top-level page, the blueprint filter is restored :slightly_smiling_face: 

https://user-images.githubusercontent.com/889383/113746901-20747380-9707-11eb-9831-9748f698b932.mp4

## Videos from all `toolbox.drillDown` calls

I have verified all the places that use `toolbox.drillDown` to make sure there are no regressions. I did not find any.

<details>
<summary>Click to expand videos</summary>

### Deploy blueprint modal (in 3 places) & Blueprints

https://user-images.githubusercontent.com/889383/113869010-8b2db980-97b0-11eb-8cae-a1124b1ad49e.mp4

### Last execution logs

https://user-images.githubusercontent.com/889383/113869085-9d0f5c80-97b0-11eb-98e3-51f46b081f20.mp4

### Agents modals (install & validate)

https://user-images.githubusercontent.com/889383/113869124-a7315b00-97b0-11eb-853a-16208cb445df.mp4

https://user-images.githubusercontent.com/889383/113869135-a8fb1e80-97b0-11eb-894b-36e4097ed396.mp4

### Deployments

https://user-images.githubusercontent.com/889383/113869195-b912fe00-97b0-11eb-95c3-4a5b6ee2b014.mp4

### Deployment wizard

I thought the component was stuck due to some JS errors, but after I come back to the browser in the end, I got to the correct drill-down page :slightly_smiling_face: 

https://user-images.githubusercontent.com/889383/113869293-d5169f80-97b0-11eb-90f7-9baa681aeb93.mp4

### Topology (drill-down to deployment node)

I could not test this functionality, because I do not know how to show a deployment node in the topology. Let me know how to get https://github.com/cloudify-cosmo/cloudify-stage/blob/master/widgets/topology/src/Topology.jsx#L314-L319 to run

</details>

## System tests

The feature will be automatically tested when doing RD-1226.

Here's a run of system tests to catch possible regressions:

https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/402/pipeline

Ran 2021-04-07 9:27 CEST

### CODE REVIEW

File: app/actions/page.ts [hunk 1]
```diff
@@ -183,32 +184,38 @@ export function changePageDescription(pageId: string, newDescription: string) {
...
         dispatch(clearWidgetsData());
 
         // Update context and location depending on page is drilldown
         if (!isDrilldown) {
             dispatch(clearContext());
+            if (drilldownContext) {
+                const newDrilldownContext: DrilldownContext[] = [{ context: drilldownContext }];
+                // eslint-disable-next-line scanjs-rules/assign_to_search
+                location.search = stringify({ c: JSON.stringify(newDrilldownContext) });
+            }
```
Comment (Gelio) [inline app/actions/page.ts:203]: This ensures that the `context` is set correctly when returning to the topmost page

File: app/actions/page.ts [hunk 2]
```diff
@@ -219,13 +226,16 @@ export function selectPage(
...
+    pageName: string,
+    context: Record<string, any>
+): ThunkAction<void, ReduxState, never, AnyAction> {
     return dispatch => {
         if (context) {
             dispatch(clearContext());
         }
         const pageId = _.snakeCase(pageName);
-        dispatch(selectPage(pageId, context, context));
+        dispatch(selectPage(pageId, false, context));
```
Comment (Gelio) [inline app/actions/page.ts:238]: Caught a bug :slightly_smiling_face:

File: backend/package.json [hunk 3]
```diff
@@ -15,7 +15,7 @@
     "db-migrate-down-to": "node migration downTo",
     "db-migrate-reset": "node migration reset",
     "db-migrate-status": "node migration status",
-    "startWithNodemon": "cross-env NODE_ENV=development nodemon --watch . --watch ../conf",
+    "startWithNodemon": "cross-env NODE_ENV=development NODE_TLS_REJECT_UNAUTHORIZED='0' nodemon --watch . --watch ../conf",
```
Comment (Gelio) [inline backend/package.json:18]: I kept setting this env variable myself, so I thought I will add it for the benefit and convenience of all of us :slightly_smiling_face:

File: widgets/deploymentsView/src/widget.tsx [hunk 4]
```diff
@@ -24,63 +24,61 @@ interface DeploymentsViewWidgetConfiguration {
     sortAscending: string;
 }
 
-// TODO(RD-1226): remove environment check
-if (process.env.NODE_ENV === 'development' || process.env.TEST) {
```
Comment (Gelio) [inline widgets/deploymentsView/src/widget.tsx:28]: Ofer asked to expose the widget in production so he can click around.

I suggest reviewing this part with _Hide whitespace changes_ on in the diff-options

image

**Comment by qooban** [comment]
I need to speed up with filters development, but I'll try to review this PR today's afternoon.

**Comment by Gelio** [comment]
> I need to speed up with filters development, but I'll try to review this PR today's afternoon.

:crying_cat_face: 

No problem :slightly_smiling_face: Looks like we may have overcommitted to deliver all the features by next Tuesday if we also do meticulous code review :worried:

File: app/actions/page.ts [hunk 5]
```diff
@@ -183,32 +184,38 @@ export function changePageDescription(pageId: string, newDescription: string) {
...
-                        context: drilldownContext || {},
-                        pageName: drilldownPageName
-                    }
-                ];
+                newDrilldownContext.push({
+                    context: drilldownContext || {},
+                    pageName: drilldownPageName
+                });
+            } else {
+                throw new Error('Drilldown page name and context not provided while doing a drilldown');
```
Comment (kubama) [inline app/actions/page.ts:218]: My understanding of this error message is that both drilldown page name and context need to be provided, while the logic above implies that only one of them is required.

File: widgets/deploymentsView/src/widget.tsx [hunk 6]
```diff
@@ -123,6 +134,17 @@ const DeploymentsView: FunctionComponent<DeploymentsViewProps> = ({ widget, tool
...
 
+    if (filteringByParentDeploymentResult.missingParentDeploymentId) {
+        const i18nMissingParentDeploymentPrefix = `${i18nMessagesPrefix}.missingParentDeploymentId`;
+
+        return (
+            <ErrorMessage
+                header={i18n.t(`${i18nMissingParentDeploymentPrefix}.header`)}
+                error={i18n.t(`${i18nMissingParentDeploymentPrefix}.message`)}
+            />
+        );
```
Comment (kubama) [inline widgets/deploymentsView/src/widget.tsx:145]: I know we already discussed this approach, but I'm still wondering - how dynamic i18n key creation would work in the context of RD-1716?

File: app/actions/page.ts (hunk 5 above)
Comment (Gelio) [reply app/actions/page.ts:218]: Hmm... Would

```suggestion
                throw new Error('Either drilldown page name or context have to be provided while doing a drilldown');
```

be better here?

File: widgets/deploymentsView/src/widget.tsx (hunk 6 above)
Comment (Gelio) [reply widgets/deploymentsView/src/widget.tsx:145]: If the concern is TS support, then TS 4.3 (beta ATM) has recently added support for operations on string literal types

https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html

See [this playground](https://www.typescriptlang.org/play?ts=4.3.0-beta#code/PTAEGUFNNALAXeAHAzgLhC+BDAxgawHsA3SAJwDMAbQgdwDpdCBbYbYAVgA4AWAZgBsATg5DgPLgHZ+PHgCh4ATyQwAUoQCWAOwA8AaQA0oAAoA+UAF5Qe0JAAe8SFoAmKUFjLaA5qAA+oLQBXZgAjclAAfjlQGJNbBydXd3hPLR9-INDwqNjQAAMAEgBvPQBfYoAiCvjHFzdjSNAq0DQm+gryouNSvOjY1q1IUjIWgKHyAG45BWUYYzIhy1AAbUHhowAGIwBGIwAmIz4jHiMOIwEjSSMuIyEdrb6Y7d3QbYPXo9eT17PXi9erq8bq87qA9ltQPQoRtlgBdWHTJQqUAAYUIWhQOgAEkYACrmKy4mqJNwLbDOdFURSgbBaRRwyKPUAACmZsFaOMhUPgrVxAEpLOZiJpnAL7LUkqyofQyK1tBRwgAlAUWIUigURUCK0ZrchMgbjMhTGbI4zYeCwTG4owAEWJdQCwTCIys2w2BJWNth9qSq0N3s1upGvJ9bkIIQAVpBcPBGbkiisbNpQPhIIpCBRQLjYQBaCKtGziknJVLpR1ZEY5XKxQolHp+UDqbT6Ixmi1W5Z6WGthbEZZe0ymJm5A3DJmlZap9OZ7OjKqI2agAAykGwpCtttD5edSzdHv73qLDr9wwDY2GoyJR6S4ajMbjsQTndAyanGazufzjc0ukMy9X646LinbdiYvYHoOoATm+M7eq0VRTCAlgWChqFoeh0xMBisYpLSKBUOaGjom4VhFEytAaM4XiQDyoBkdWsTMJAKAoNg1HoHRw4MTSrT0dx-GwKuzjkK0ADk2AhLgokGFx-ExExLFsZAYkoCwMCEBa5CibJ1alDJcmxCEvE6QJQkiaAomOFg2kGTEpQ6fZuT2aUNJuFhWDGkiMC4mQeEEfARFaG2lpLCua7MToXnvrhGL+YFKCmMaFCBFoMaBaA8DMkg5rslmvmxYR6LBSgAp8TESG4gA8jalVyPZcjubGCmsex8yQBQGh2EsokUVRNH0M1SkoKJUyNTSWJmS6GXMrWg2tQsHV2KU9DYPQgnkuQeR8qNxGxtgACyzEtTAViZbNR1DW1i3Latc2QFtO3YaAIQTRtU1ncUd0oFdnXLSEa2TQ9chIQAQpANC0DSCzxCoMaQM4thkGQhBkCgDW7c9ABy6IAKJ2BoWBOPAh2KdRSwfUUX0-Ut9D-XdQNjUihDkzNn0XfN7W-fQKDOBQrG8wNKBeA9QA) for an example usage.

File: app/actions/page.ts (hunk 5 above)
Comment (kubama) [reply app/actions/page.ts:218]: Clear to me now 👍.

**Comment by Gelio** [comment]
In https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/402/pipeline/, the system tests all passed, but the unit tests failed. I will rebase the PR on `master` (since the base branch was merged), and only fix the unit tests. I will not rerun the system tests because the application code won't change :slightly_smiling_face:

File: widgets/deploymentsView/src/widget.tsx (hunk 6 above)
Comment (kubama) [reply widgets/deploymentsView/src/widget.tsx:145]: Cool, good to know, all good in this case then 👍.

Side question: what if we used function to create i18n key (I've seen such approach in another PR)? Would it still work with i18n TS support enabled?

Comment (Gelio) [reply widgets/deploymentsView/src/widget.tsx:145]: > Side question: what if we used function to create i18n key (I've seen such approach in another PR)? Would it still work with i18n TS support enabled?

Yes :smile: Here's a [TS playground with some more magic](https://www.typescriptlang.org/play?ts=4.3.0-beta#code/PTAEGUFNNALAXeAHAzgLhC+BDAxgawHsA3SAJwDMAbQgdwDpdCBbYbYAVgA4AWAZgBsATg5DgPLgHZ+PHgCh4ATyQwAUoQCWAOwA8AaQA0oAAoA+UAF5Qe0JAAe8SFoAmKUFjLaA5qAA+oLQBXZgAjclAAfjlQGJNbBydXd3hPLR9-INDwqNjQAAMAEgBvPQBfYoAiCvjHFzdjSNAq0DQm+gryouNSvOjY1q1IUjIWgKHyAG45BWUYYzIhy1AAbUHhowAGIwBGIwAmIz4jHiMOIwEjSSMuIyEdrb6Y7d3QbYPXo9eT17PXi9erq8bq87qA9ltQPQoRtlgBdWHTJQqUAAYUIWhQOgAEkYACrmKy4mqJNwLbDOdFURSgbBaRRwyKPUAACmZsFaOMhUPgrVxAEpLOZiJpnAL7LUkqyofQyK1tBRwgAlAUWIUigURUCK0ZrchMgbjMhTGbI4zYeCwTG4owAEWJdQCwTCIys2w2BJWNth9qSq0N3s1upGvJ9bkIIQAVpBcPBGbkiisbNpQPhIIpCBRQLjYQBaCKtGziknJVLpR1ZEY5XKxQolHp+UDqbT6Ixmi1W5Z6WGthbEZZe0ymJm5A3DJmlZap9OZ7OjKqI2agAAykGwpCtttD5edSzdHv73qLDr9wwDY2GoyJR6S4ajMbjsQTndAyanGazufzjc0ukMy9X646LinbdiYvYHoOoATm+M7eq0VRTCAlgWChqFoeh0xMBisYpLSKBUOaGjom4VhFEytAaM4XiQDyoBkdWsTMJAKAoNg1HoHRw4MTSrT0dx-GwKuzjkK0ADk2AhLgokGFx-ExExLFsZAYkoCwMCEBa5CibJ1alDJcmxCEvE6QJQkiaAomOFg2kGTEpQ6fZuT2aUNJuFhWDGkiMC4mQeEEfARFaG2lpLCua7MToXnvrhGL+YFKCmMaFCBFoMaBaA8DMkg5rslmvmxYR6LBSgAp8TESG4gA8jalVyPZmHEbGCmsex8yQBQGh2EsokUVRNH0M1SkoKJUzubG2BYmZLoZcytaDa1CwdXYpT0Ng9CCeS5B5Hyo2NTSACyzEtTAViZXNR1DW1S0rWt82QNtu3YaAISTZt01ncUd0oFdnUrSE61TQ9chIQAQpANC0DSCzxCoMaQM4thkGQhBkCgchjc9ABy6IAKJ2BoWBOPAh2KdRSwfUUX0-ct9D-XdQMY0ihDk7Nn0XQt7W-fQKDOBQrG8wNKBeEDwNgAAqig3hwODKhkNMyWpQF6KgNlFqTVQsvMiggQUEtrRhYBUWZjF+GFRiyw9ZR1HwKJsIW19tumKVTILPAgRkFoM15L11sDezzH0MU2u679QP2RjxDYFQlHgDrS1LKrsDq5r4kA29ok7eje3aJH0fOLHIddVYifJ+QzKp1o6I5vYBMBWkGcTEAA) (see the part at the bottom)

Comment (kubama) [reply widgets/deploymentsView/src/widget.tsx:145]: Cool, thanks for explaining.

**Comment by Gelio** [comment]
I've only rebased the PR on master so far. No code changes yet

File: app/actions/page.ts (hunk 5 above)
Comment (Gelio) [reply app/actions/page.ts]: Great :+1: Done

**Comment by Gelio** [comment]
I've fixed the tests in the latest commit. `pagesReducer` tests used the drill-down functionality in an artificial way and are white-box by testing the exact sequence of dispatched actions (which was changed with this PR) :slightly_smiling_face:

**Comment by Gelio** [comment]
I see that the size check is failing in CI. I will wait for #1284 to be merged, as it will also have to change the `.size-limit` file, and by waiting I will avoid resolving conflicts :slightly_smiling_face:

**Review by qooban — COMMENTED** [review]
I didn't go very deep in the review, just quickly went over the code, didn't focus on the style. Don't see anything problematic.

I'm more interested in your tests for backward compatibility as you are changing core functionality. Have you tested manually all the drilldown cases in the application?
Quick search gives me the following:
image

If all things drilldown-related like context setting in redux, browser navigation (history back, forward) and breadcrumbs work fine, then I'm going to approve this PR without getting into more details.

**Comment by Gelio** [comment]
> Have you tested manually all the drilldown cases in the application?

I didn't :smile: I thought that the system tests have it covered. Since you asked, I'll verify all those places and post the results soon :slightly_smiling_face:

**Comment by qooban** [comment]
> I didn't smile I thought that the system tests have it covered. Since you asked, I'll verify all those places and post the results soon slightly_smiling_face

Probably some drilldown features are covered by jest/cypress tests, but I don't think we have good tests for drilldown pages functionality. So I think it is worth, at this stage, at least to check that manually. Of course, in a perfect world, we would have better automated tests for that 🙂

**Comment by Gelio** [comment]
> So I think it is worth, at this stage, at least to check that manually

Yeah, yeah, I agree :slightly_smiling_face: 

I have added a _Videos from all toolbox.drillDown calls_ section in the description of the PR.

I have tested all the places I could find (all that you listed and additionally the deployment wizard buttons widget - I was grepping for `drillDown(` :slightly_smiling_face:), except for the Topology - I could not use the drill-down functionality there, because I don't know how to show deployment nodes in the topology. I commented on that in the PR description :slightly_smiling_face: 

All other places seem to work fine, as far as I have checked (going back & forth in the browser's history, and going back using the breadcrumbs) :slightly_smiling_face:

**Review by qooban — APPROVED** [review]
Thanks for all those tests, I really appreciate that 👍🏻 Good work!

**Review by kubama — APPROVED** [review]

**Comment by Gelio** [comment]
Thanks for the reviews :slightly_smiling_face: I'll merge the PR without rerunning the system tests (the reason is mentioned in https://github.com/cloudify-cosmo/cloudify-stage/pull/1290#issuecomment-814741918) to avoid taking up CI time that provides little value

**Pull request merged**